common/src/main/java/org/apache/nemo/common/ir/IRDAG.java

```diff
@@ -151,73 +176,153 @@ public void insert(final StreamVertex streamVertex, final IREdge edgeToStreamize
-
-          // Clone the edgeToGetStatisticsOf
-          final IREdge clone = new IREdge(CommunicationPatternProperty.Value.OneToOne, edge.getSrc(), mbv);
-          clone.setProperty(EncoderProperty.of(edge.getPropertyValue(EncoderProperty.class).get()));
-          clone.setProperty(DecoderProperty.of(edge.getPropertyValue(DecoderProperty.class).get()));
-          edge.getPropertyValue(AdditionalOutputTagProperty.class).ifPresent(tag -> {
-            clone.setProperty(AdditionalOutputTagProperty.of(tag));
-          });
-          builder.connectVertices(clone);
-
-          // messageBarrierVertex to the messageAggregatorVertex
-          final IREdge edgeToABV = edgeBetweenMessageVertices(mbv,
-            messageAggregatorVertex, mbvOutputEncoder, mbvOutputDecoder, currentMetricCollectionId);
-          builder.connectVertices(edgeToABV);
-
-          // The original edge
-          // We then insert the vertex with MessageBarrierTransform and vertex with MessageAggregatorTransform
-          // between the vertex and incoming vertices.
-          final IREdge edgeToOriginalDst =
-            new IREdge(edge.getPropertyValue(CommunicationPatternProperty.class).get(), edge.getSrc(), v);
-          edge.copyExecutionPropertiesTo(edgeToOriginalDst);
-          edgeToOriginalDst.setPropertyPermanently(MessageIdProperty.of(currentMetricCollectionId));
-          builder.connectVertices(edgeToOriginalDst);
-        } else {
-          // NO MATCH, so simply connect vertices as before.
-          builder.connectVertices(edge);
+      modifiedDAG.getIncomingEdgesOf(v).forEach(inEdge -> {
+        if (edgesToOptimize.contains(inEdge)) {
+          inEdge.setPropertyPermanently(MessageIdProperty.of(messageAggregatorVertex.getMessageId()));
         }
-      }
+      });
+    });
+
+    modifiedDAG = builder.build(); // update the DAG.
+  }
+
+  /**
+   * Inserts a set of samplingVertices that process sampled data.
+   *
+   * This method automatically inserts the following three types of edges.
+   * (1) Edges between samplingVertices to reflect the original relationship
+   * (2) Edges from the original IRDAG to samplingVertices that clone the inEdges of the original vertices
+   * (3) Edges from the samplingVertices to the original IRDAG to respect executeAfterSamplingVertices
+   *
+   * Suppose the caller supplies the following arguments to perform a "sampled run" of vertices {V1, V2},
+   * prior to executing them.
+   * - samplingVertices: {V1', V2'}
+   * - childrenOfSamplingVertices: {V1}
+   *
+   * Before: V1 - oneToOneEdge - V2 - shuffleEdge - V3
+   * After: V1' - oneToOneEdge - V2' - controlEdge - V1 - oneToOneEdge - V2 - shuffleEdge - V3
+   *
+   * This preserves semantics as the original IRDAG remains unchanged and unaffected.
+   *
+   * @param samplingVertices to insert.
+   * @param executeAfterSamplingVertices that must be executed after samplingVertices.
+   */
+  public void insert(final Set<SamplingVertex> samplingVertices,
```

Comment:
Why don't we get the `Set` of the original vertices to sample instead of already sampled vertices?
If we receive the sampled vertices, opt pass builders can give some already connected sampling vertices.